### PR TITLE
Update logic for identifying and adding missing items in repository changes

### PR DESCRIPTION
**Pull request by UdiBen**
### Background

Object list shows wrong date when hitting the changes pagination window 

### Bug Fix

The root cause is that changes and objects are paginated together, but changes might be "ahead" and include items that were not retrieved from objects list. Then there's a migrate function that handles each change with no object as deleted object.

The fix: only consider change of type deletion/prefix change as actually missing items.

### CODE REVIEW

File: webui/.eslintrc [hunk 1]
```diff
@@ -1,35 +1,32 @@
 {
```
Comment (UdiBen) [inline webui/.eslintrc:1]: Changed only because prettier executed

File: webui/src/lib/components/repository/mergeResults.test.ts [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+import { mergeResults } from './mergeResults';
+import { Entry, ChangesData, EntryWithDiff } from './types';
+
+describe('mergeResults', () => {
+    describe('basic behavior', () => {
+        it('returns null when results is null', () => {
+            expect(mergeResults(null, { results: [] }, false)).toEqual([]);
+        });
+
+        it('returns undefined when results is undefined', () => {
```
Comment (arielshaqed) [inline webui/src/lib/components/repository/mergeResults.test.ts]: Results is not undefined here.  I think "results" refers to two things, so perhaps use a different name.

File: webui/src/lib/components/repository/mergeResults.test.ts [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+
+            const merged = mergeResults(results, {} as ChangesData, false);
+
+            expect(merged).toEqual([
+                { path: 'file1.txt', path_type: 'object' },
+                { path: 'file2.txt', path_type: 'object' },
+            ]);
+        });
+
+        it('returns sorted results when showChangesOnly is true', () => {
```
Comment (arielshaqed) [inline webui/src/lib/components/repository/mergeResults.test.ts]: - Why not test when `showChangesOnly === false`?
- Why not parametrize over `[true, false]` (or run a loop) for _all_ tests which needs to be tested for both values?

File: webui/src/lib/components/repository/mergeResults.test.ts [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+                    { path: 'dir/', type: 'removed', path_type: 'common_prefix' },
+                    { path: 'dir/file.txt', type: 'removed', path_type: 'object' },
+                ],
+            };
+
+            const merged = mergeResults(results, changesData, false);
+
+            // Directory should keep 'removed' type, not be overridden by 'changed'
+            expect(merged.find((r) => r.path === 'dir/')).toMatchObject({ diff_type: 'removed' });
+        });
```
Comment (arielshaqed) [inline webui/src/lib/components/repository/mergeResults.test.ts:154]: I think we're missing some mixed-type cases:

- What happens if a path (not a prefix!) `a/b/c` is replaced by _added_ `a/b/c/d`, `a/b/c/e`?

  ```
  Before: a/b/c (not a prefix!)
  After:  a/b/c/d
          a/b/c/e
  ```
- What happens if a path (not a prefix!) `a/b/c` has subpaths _added_ below it?

  ```
  Before: a/b/c (not a prefix!)
  After:  a/b/c
          a/b/c/d
          a/b/c/e
  ```
And similar...

File: webui/src/lib/components/repository/mergeResults.test.ts [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+
+            const merged = mergeResults(results, changesData, false);
+
+            expect(merged).toHaveLength(2);
+            expect(merged.find((r) => r.path === 'deleted.txt')).toMatchObject({
+                path: 'deleted.txt',
+                type: 'removed',
+                path_type: 'object',
+                diff_type: 'removed',
+            });
```
Comment (arielshaqed) [inline webui/src/lib/components/repository/mergeResults.test.ts:184]: Can't we use an array-aware deep comparator?  Something like `expect(merged).toEqual(whateverWeExpect)` should be easier to read and write, and will probably give better errors messages when it fails!

File: webui/src/pages/repositories/repository/objects.jsx [hunk 6]
```diff
@@ -967,73 +968,7 @@ const TreeContainer = ({
...
-                return {
-                    ...item,
-                    diff_type: mappedType,
-                };
-            }),
-        ];
-
-        // Sort to maintain proper order
-        return allResults.sort((a, b) => a.path.localeCompare(b.path));
+        return mergeResults(results, changesData, showChangesOnly);
```
Comment (arielshaqed) [inline webui/src/pages/repositories/repository/objects.jsx]: Nit: call the function directly, it's an expression.
```suggestion
const mergedResults = React.useMemo(() => mergeResults(results, changesData, showChangesOnly), [results, changesData, showChangesOnly]);
```

**Review by arielshaqed — APPROVED** [review]
Sorry; I had one job.

glowing LGTM button

File: webui/src/lib/components/repository/mergeResults.property.test.ts [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+import { Entry, ChangesData, DiffType } from './types';
+
+/**
+ * Property-based tests for mergeResults function using fast-check.
+ *
+ * These tests complement the example-based tests by verifying invariants
+ * across thousands of randomly generated inputs to catch edge cases.
+ */
+
+// ==================== Arbitraries (Random Data Generators) ====================
```
Comment (arielshaqed) [inline webui/src/lib/components/repository/mergeResults.property.test.ts]: Please remove aligned title lines: they are hard to maintain (need to fix line length whenever the text changes).  Either remove it entirely or remove the header.

```suggestion
// ** Random Data Generators
```

If you prefer then mention "arbitraries" in a link - it seems to be a fuzzer term rather than a property-based testing term.

```suggestion
// ** Random data generators.  See
// https://fast-check.dev/docs/introduction/getting-started/#arbitrary.
``1

File: webui/src/lib/components/repository/mergeResults.property.test.ts [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+
+/**
+ * Generates valid path segments (alphanumeric + common chars)
```
Comment (arielshaqed) [inline webui/src/lib/components/repository/mergeResults.property.test.ts]: Personally I prefer "returns ..." here, and use the term "generates" here to a function* that `yield`s values (aka a "[generator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator)").  But this is an extreme nitpick: most of our webui functions don't even _have_ a JSDoc...

File: webui/src/lib/components/repository/mergeResults.property.test.ts [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+                    const changesData: ChangesData = { results: changes };
+                    const merged = mergeResults(results, changesData, false);
+
+                    // Verify sorting
+                    for (let i = 0; i < merged.length - 1; i++) {
+                        const comparison = merged[i].path.localeCompare(merged[i + 1].path);
+                        expect(comparison).toBeLessThanOrEqual(0);
+                    }
+                }),
+                { numRuns: 100 },
```
Comment (arielshaqed) [inline webui/src/lib/components/repository/mergeResults.property.test.ts]:
```suggestion
                { numRuns: 2000 },
```
(PR description mentions "thousands"; unless it's too slow I say the description is correct).

File: webui/src/lib/components/repository/mergeResults.property.test.ts [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+            );
+        });
+
+        it('PROPERTY: no duplicate paths in output', () => {
+            fc.assert(
+                fc.property(uniqueEntriesArbitrary, uniqueChangesArbitrary, (results, changes) => {
```
Comment (arielshaqed) [inline webui/src/lib/components/repository/mergeResults.property.test.ts]: I am not sure that these sources can meaningfully check this property with fast-check.  I would prefer a (much) higher proportion of duplicate paths.  So generate _three_ sets of unique changes $A, B, C$, from these generate $A\cup C, B\cup C$, shuffle each of these two sets, and pass _those_.

File: webui/src/lib/components/repository/mergeResults.property.test.ts [hunk 11]
```diff
@@ -0,0 +1,418 @@
...
+                    const changesData: ChangesData = { results: changes };
+                    const merged = mergeResults(results, changesData, false);
+
+                    const mergedPaths = new Set(merged.map((e) => e.path));
+
+                    // Every result should appear in merged output
+                    results.forEach((entry) => {
+                        expect(mergedPaths.has(entry.path)).toBe(true);
+                    });
```
Comment (arielshaqed) [inline webui/src/lib/components/repository/mergeResults.property.test.ts]: Ideally also verify sizes are equal (before calling the Set constructor), or we risk dupes or even random entries.  Again, can we just check for arrays with equal paths after sorting (or arrays with equal paths in possibly arbitrary order)?
(Not repeating this comment, it might well apply in other places)

File: webui/src/lib/components/repository/mergeResults.property.test.ts [hunk 12]
```diff
@@ -0,0 +1,418 @@
...
+        it('PROPERTY: no duplicate paths in output', () => {
+            fc.assert(
+                fc.property(uniqueEntriesArbitrary, uniqueChangesArbitrary, (results, changes) => {
+                    const changesData: ChangesData = { results: changes };
+                    const merged = mergeResults(results, changesData, false);
+
+                    const paths = merged.map((e) => e.path);
+                    const uniquePaths = new Set(paths);
+
+                    expect(paths.length).toBe(uniquePaths.size);
```
Comment (arielshaqed) [inline webui/src/lib/components/repository/mergeResults.property.test.ts]: This cannot detect if `merged` has any duplicate paths.   Can we just check for arrays with equal paths after sorting (or arrays with equal paths in possibly arbitrary order)?

File: webui/src/lib/components/repository/mergeResults.property.test.ts [hunk 13]
```diff
@@ -0,0 +1,418 @@
...
+                    // Merged results should contain at least all original results
+                    expect(merged.length).toBeGreaterThanOrEqual(results.length);
+                }),
+                { numRuns: 100 },
+            );
+        });
+
+        it('PROPERTY: properties are preserved when entries are merged', () => {
+            fc.assert(
+                fc.property(uniqueEntriesArbitrary, uniqueChangesArbitrary, (results, changes) => {
```
Comment (arielshaqed) [inline webui/src/lib/components/repository/mergeResults.property.test.ts]: Again, here I expect  almost no merges.  I think we need the $A, B, C$ approach from above.

File: webui/src/lib/components/repository/mergeResults.test.ts [hunk 14]
```diff
@@ -331,6 +335,142 @@ describe('mergeResults', () => {
...
+        it('handles file with subpaths added below it', () => {
+            // Before: a/b/c (file/object)
+            // After:  a/b/c (file/object - still exists)
+            //         a/b/c/d (file)
+            //         a/b/c/e (file)
+            const results: Entry[] = [
+                { path: 'a/b/c', path_type: 'object' },
+                { path: 'a/b/c/d', path_type: 'object' },
+                { path: 'a/b/c/e', path_type: 'object' },
+            ];
```
Comment (arielshaqed) [inline webui/src/lib/components/repository/mergeResults.test.ts]: Again, this describes the after rather than the before.

File: webui/src/lib/components/repository/mergeResults.test.ts [hunk 15]
```diff
@@ -331,6 +335,142 @@ describe('mergeResults', () => {
...
+            //         a/b/c/e (file)
+            const results: Entry[] = [
+                { path: 'a/b/c/', path_type: 'common_prefix' },
+                { path: 'a/b/c/d', path_type: 'object' },
+                { path: 'a/b/c/e', path_type: 'object' },
+            ];
+            const changesData: ChangesData = {
+                results: [
+                    { path: 'a/b/c', type: 'removed', path_type: 'object' }, // Old file removed
+                    { path: 'a/b/c/', type: 'added', path_type: 'common_prefix' }, // New directory added
```
Comment (arielshaqed) [inline webui/src/lib/components/repository/mergeResults.test.ts]: Does the lakefs diff API return both `a/b/c/` and `a/b/c/d` and `a/b/c/d/e` here?  Strange, I thought you only get the prefix in this case...  (This does not indicate a bug in the code, of course - merely that we might be testing the wrong thing!  So please verify that these actions do indeed cause these changes to show up...)

Comment (UdiBen) [reply webui/src/lib/components/repository/mergeResults.test.ts]: You're right. It only returns the prefix. 
I will fix the tests. 
Just a nit picking question - I think the actual semantic of the API is not truly the concern of the mergeResults util (in other words - if this semantic of the API ever changes, we will need to update the tests?)

Comment (UdiBen) [reply webui/src/lib/components/repository/mergeResults.test.ts]: Scratch that - I fixed the tests to reflect the actual API behavior. It got me to dig into both `ls` and `diff` behavior.
Thanks
